streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.kafka.streams.kstream.Consumed;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.internals.StateDirectory;
+import org.apache.kafka.streams.state.internals.OffsetCheckpoint;
+import org.apache.kafka.test.TestUtils;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Properties;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.kafka.test.TestUtils.waitForCondition;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * An integration test to verify the conversion of a dirty-closed EOS
+ * task towards a standby task is safe across restarts of the application.
+ */
+public class StandbyTaskEOSIntegrationTest {
+
+    private final String inputTopic = "input";
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(3);
+
+    @Before
+    public void createTopics() throws Exception {
+        CLUSTER.createTopic(inputTopic, 1, 3);
+    }
+
+    @Test
+    public void surviveWithOneTaskAsStandby() throws ExecutionException, InterruptedException, IOException {
+        IntegrationTestUtils.produceKeyValuesSynchronouslyWithTimestamp(
+            inputTopic,
+            Collections.singletonList(
+                new KeyValue<>(0, 0)),
+            TestUtils.producerConfig(
+                CLUSTER.bootstrapServers(),
+                IntegerSerializer.class,
+                IntegerSerializer.class,
+                new Properties()),
+            10L);
+
+        final String appId = "eos-test-app";
+        final String stateDirPath = TestUtils.tempDirectory(appId).getPath();
+
+        final CountDownLatch instanceLatch = new CountDownLatch(1);
+
+        final String stateDirPathOne = stateDirPath + "/" + appId + "-1/";
+        final KafkaStreams streamInstanceOne =
+            buildStreamWithDirtyStateDir(appId, stateDirPathOne, instanceLatch);
+
+        final String stateDirPathTwo = stateDirPath + "/" + appId + "-2/";
+        final KafkaStreams streamInstanceTwo =
+            buildStreamWithDirtyStateDir(appId, stateDirPathTwo, instanceLatch);
+
+        streamInstanceOne.start();
+
+        streamInstanceTwo.start();
+
+        // Wait for the record to be processed
+        assertTrue(instanceLatch.await(15, TimeUnit.SECONDS));
+
+        waitForCondition(() -> streamInstanceOne.state().equals(KafkaStreams.State.RUNNING),
+            "Stream instance one should be up and running by now");
+        waitForCondition(() -> streamInstanceTwo.state().equals(KafkaStreams.State.RUNNING),
+            "Stream instance one should be up and running by now");
+
+        streamInstanceOne.close(Duration.ofSeconds(30));
```

Comment:
For my own education: before the fix, this integration test will fail when instance-2 is started?

Reply:
Yes, actually either instance-1 or instance-2 would fail, depending on which box gets standby assignment. There would be a IllegalState + NPE exception sequence happening.